File: frontend-neopop/src/parsers/generic.js
/**
 * Generic credit card statement parser.
 * Direct port of backend/parsers/generic.py (173 lines)
 *
 * Fallback parser for banks without a dedicated parser.
 */

import { parseNumericDate } from './base.js';

/**
 * @param {string} bank
 * @param {string[]} allLines
 * @param {string} fullText
 * @returns {import('./base.js').ParsedStatement}
 */
export function parseGeneric(bank, allLines, fullText) {
  const [periodStart, periodEnd] = extractPeriod(fullText);
  const cardLast4 = extractCardLast4(fullText);
  const totalAmountDue = extractTotalAmountDue(fullText);
  const creditLimit = extractCreditLimit(fullText);
  const transactions = extractTransactions(allLines);

  return {
    bank,
    period_start: periodStart,
    period_end: periodEnd,
    transactions,
    card_last4: cardLast4,
    total_amount_due: totalAmountDue,
    credit_limit: creditLimit,
  };
}

function extractPeriod(text) {
  const patterns = [
    /Statement\s+Period[:\s]+(\d{2}[/-]\d{2}[/-]\d{2,4})\s*[-–to]+\s*(\d{2}[/-]\d{2}[/-]\d{2,4})/i,
    /Billing\s+Period[:\s]+(\d{2}[/-]\d{2}[/-]\d{2,4})\s*[-–to]+\s*(\d{2}[/-]\d{2}[/-]\d{2,4})/i,
    /From\s+(\d{2}[/-]\d{2}[/-]\d{2,4})\s+[Tt]o\s+(\d{2}[/-]\d{2}[/-]\d{2,4})/i,
    /(\d{2}[/-]\d{2}[/-]\d{4})\s*[-–]\s*(\d{2}[/-]\d{2}[/-]\d{4})/,
  ];

  for (const pat of patterns) {
    const m = pat.exec(text);
    if (m) {
      const start = parseNumericDate(m[1]);
      const end = parseNumericDate(m[2]);
      if (start && end) return [start, end];
    }
  }

  return [null, null];
}

function extractCardLast4(text) {
  const m = /(\d{4,6})[Xx*]{4,}(\d{4})/.exec(text);
  return m ? m[2] : null;
}

function extractTotalAmountDue(text) {
  const patterns = [
    /Total\s+Amount\s+Due.*?([\d,]+\.\d{2})/is,
    /Total\s+Payment\s+Due.*?([\d,]+\.\d{2})/is,
    /Amount\s+Payable.*?([\d,]+\.\d{2})/is,
  ];
  for (const pat of patterns) {
    const m = pat.exec(text);
    if (m) {
      try { return parseFloat(m[1].replace(/,/g, '')); } catch { /* ignore */ }
    }
  }
  return null;
}

function extractCreditLimit(text) {
  const patterns = [
    /Credit\s+Limit.*?([\d,]+\.\d{2})/is,
    /Total\s+Credit\s+Limit.*?([\d,]+\.\d{2})/is,
  ];
  for (const pat of patterns) {
    const m = pat.exec(text);
    if (m) {
      try {
        const val = parseFloat(m[1].replace(/,/g, ''));
        if (val > 0) return val;
      } catch { /* ignore */ }
    }
  }
  return null;
}

function extractTransactions(lines) {
  const transactions = [];

  for (let rawLine of lines) {
    let line = rawLine.trim();
    if (!line) continue;
    line = line.replace(/\s+/g, ' ').trim();

    const tx = parseTransactionLine(line);
    if (tx) transactions.push(tx);
  }

  return transactions;
}

function parseTransactionLine(line) {
  const m = /^(\d{2}[/-]\d{2}[/-]\d{4})\s+(.+?)\s+([\d,]+\.\d{2})\s*(Dr|Cr|CR|DR)?\s*$/i.exec(line);
  if (!m) return null;

  const parsedDate = parseNumericDate(m[1]);
  if (!parsedDate) return null;

  const rawDesc = m[2].trim();
  const amountStr = m[3].replace(/,/g, '');
  const direction = (m[4] || '').trim().toLowerCase();

  const amount = parseFloat(amountStr);
  if (isNaN(amount) || amount <= 0) return null;

  const isCredit = direction === 'cr';

  return {
    date: parsedDate,
    merchant: rawDesc ? rawDesc.slice(0, 512) : 'Unknown',
    amount,
    type: isCredit ? 'credit' : 'debit',
    description: rawDesc,
  };
}

File: backend/parsers/generic.py
"""Generic credit card statement parser.

Fallback parser for banks without a dedicated parser. Uses common
patterns found in Indian bank credit card statements to extract
statement metadata and transactions.
"""

import logging
import re
from datetime import date, datetime
from typing import List, Optional, Tuple

import pdfplumber

from backend.parsers.base import BaseParser, ParsedStatement, ParsedTransaction
from backend.parsers.currency_infer import infer_currency_from_document_text

logger = logging.getLogger(__name__)


class GenericParser(BaseParser):
    """Fallback parser for unsupported bank statements."""

    def __init__(self, bank: str = "unknown"):
        self._bank = bank

    def parse(self, pdf_path: str) -> ParsedStatement:
        all_lines: List[str] = []
        full_text = ""

        with pdfplumber.open(pdf_path) as pdf:
            for page in pdf.pages:
                text = page.extract_text() or ""
                full_text += text + "\n"
                all_lines.extend(text.split("\n"))

        period_start, period_end = self._extract_period(full_text)
        card_last4 = self._extract_card_last4(full_text)
        total_amount_due = self._extract_total_amount_due(full_text)
        credit_limit = self._extract_credit_limit(full_text)
        transactions = self._extract_transactions(all_lines)
        currency = infer_currency_from_document_text(full_text)

        logger.info(
            "Generic parse (%s): card=%s period=%s..%s txns=%d due=%s limit=%s currency=%s",
            self._bank, card_last4, period_start, period_end,
            len(transactions), total_amount_due, credit_limit, currency,
        )

        return ParsedStatement(
            bank=self._bank,
            period_start=period_start,
            period_end=period_end,
            transactions=transactions,
            card_last4=card_last4,
            total_amount_due=total_amount_due,
            credit_limit=credit_limit,
            currency=currency,
        )

    def _extract_period(self, text: str) -> Tuple[Optional[date], Optional[date]]:
        patterns = [
            r"Statement\s+Period[:\s]+(\d{2}[/-]\d{2}[/-]\d{2,4})\s*[-–to]+\s*(\d{2}[/-]\d{2}[/-]\d{2,4})",
            r"Billing\s+Period[:\s]+(\d{2}[/-]\d{2}[/-]\d{2,4})\s*[-–to]+\s*(\d{2}[/-]\d{2}[/-]\d{2,4})",
            r"From\s+(\d{2}[/-]\d{2}[/-]\d{2,4})\s+[Tt]o\s+(\d{2}[/-]\d{2}[/-]\d{2,4})",
            r"(\d{2}[/-]\d{2}[/-]\d{4})\s*[-–]\s*(\d{2}[/-]\d{2}[/-]\d{4})",
        ]
        for pat in patterns:
            m = re.search(pat, text, re.IGNORECASE)
            if m:
                start = self._parse_date(m.group(1))
                end = self._parse_date(m.group(2))
                if start and end:
                    return start, end
        return None, None

    @staticmethod
    def _extract_card_last4(text: str) -> Optional[str]:
        for m in re.finditer(r"(\d{4,6})[Xx*]{4,}(\d{4})", text):
            return m.group(2)
        return None

    @staticmethod
    def _extract_total_amount_due(text: str) -> Optional[float]:
        patterns = [
            r"Total\s+Amount\s+Due.*?([\d,]+\.\d{2})",
            r"Total\s+Payment\s+Due.*?([\d,]+\.\d{2})",
            r"Amount\s+Payable.*?([\d,]+\.\d{2})",
        ]
        for pat in patterns:
            m = re.search(pat, text, re.IGNORECASE | re.DOTALL)
            if m:
                try:
                    return float(m.group(1).replace(",", ""))
                except ValueError:
                    pass
        return None

    @staticmethod
    def _extract_credit_limit(text: str) -> Optional[float]:
        patterns = [
            r"Credit\s+Limit.*?([\d,]+\.\d{2})",
            r"Total\s+Credit\s+Limit.*?([\d,]+\.\d{2})",
        ]
        for pat in patterns:
            m = re.search(pat, text, re.IGNORECASE | re.DOTALL)
            if m:
                try:
                    val = float(m.group(1).replace(",", ""))
                    if val > 0:
                        return val
                except ValueError:
                    pass
        return None

    def _extract_transactions(self, lines: List[str]) -> List[ParsedTransaction]:
        transactions: List[ParsedTransaction] = []

        for raw_line in lines:
            line = raw_line.strip()
            if not line:
                continue
            line = re.sub(r"\s+", " ", line).strip()

            tx = self._parse_transaction_line(line)
            if tx:
                transactions.append(tx)

        return transactions

    def _parse_transaction_line(self, line: str) -> Optional[ParsedTransaction]:
        # Pattern: DD/MM/YYYY description amount [Dr/Cr]
        m = re.match(
            r"^(\d{2}[/-]\d{2}[/-]\d{4})\s+(.+?)\s+([\d,]+\.\d{2})\s*(Dr|Cr|CR|DR)?\s*$",
            line,
            re.IGNORECASE,
        )
        if not m:
            return None

        parsed_date = self._parse_date(m.group(1))
        if not parsed_date:
            return None

        raw_desc = m.group(2).strip()
        amount_str = m.group(3).replace(",", "")
        direction = (m.group(4) or "").strip().lower()

        try:
            amount = float(amount_str)
        except ValueError:
            return None

        if amount <= 0:
            return None

        is_credit = direction in ("cr",)
        tx_type = "credit" if is_credit else "debit"

        return ParsedTransaction(
            date=parsed_date,
            merchant=raw_desc[:512] if raw_desc else "Unknown",
            amount=amount,
            type=tx_type,
            description=raw_desc,
        )

    @staticmethod
    def _parse_date(date_str: str) -> Optional[date]:
        for fmt in ("%d/%m/%Y", "%d-%m-%Y", "%d/%m/%y", "%d-%m-%y"):
            try:
                return datetime.strptime(date_str.strip(), fmt).date()
            except ValueError:
                continue
        return None
